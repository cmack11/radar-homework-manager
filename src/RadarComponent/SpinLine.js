import React, { Component } from 'react';
import util from './utils.js'

class SpinLine extends Component {


	constructor(props) {
		super();
		let state = {angle:(Math.PI/2)}
		this.state = state;
	}

	componentDidMount() {
		let interval  = 50;
		let rpm = this.props.rpm;
		let step = rpm * (360 / (60000 / interval));
		setInterval(() => {
			let state = this.state;
			state.angle = this.state.angle + step;
			if(this.state.angle > 360)
				state.angle = 0;
			this.setState(state);
		},interval)
	}

	
	render() {

		let edge = util.polarToCartesian(this.props.center.x, this.props.center.y, this.props.radius, this.state.angle)
		return (
			<line 
			id="spinLine"
			stroke={this.props.lineColor}
			strokeWidth="10"
			x1={this.props.center.x} y1={this.props.center.y}
			x2={edge.x} y2={edge.y}
			/>
				

		)
	}
}

export default SpinLine;