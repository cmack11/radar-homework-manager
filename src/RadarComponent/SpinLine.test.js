import React from 'react';
import SpinLine from './SpinLine.js';
import { shallow, mount, render } from 'enzyme';
import * as Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

beforeAll(() => {
  Enzyme.configure({ adapter: new Adapter() })
  jest.useFakeTimers();
});

const line = {
  center:{x:10,y:15},
  length:100,
}

const setLineAngle = function() {

}

describe('initialization', () => {

  it('should initialize the SpinLine', () => {
    const wrapper = shallow(
      <SpinLine show setLineAngle={setLineAngle} rpm="6" radius={line.length} center={line.center} />,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should initialize a null SpinLine', () => {
    const wrapper = shallow(
      <SpinLine setLineAngle={setLineAngle} rpm="6" radius={line.length} center={line.center} />,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should set an interval to update the coordinates of the spinLine', () => {
    const wrapper = shallow(
      <SpinLine show setLineAngle={setLineAngle} rpm="6" radius={line.length} center={line.center} />,
    );

    expect(wrapper.instance().spinInterval).toBeDefined();
    expect(setInterval).toBeCalled();
    expect(setInterval).toBeCalledWith(expect.any(Function),50);

  });

  it('should fill the setLineAngle props function', () => {
    const wrapper = shallow(
      <SpinLine show setLineAngle={setLineAngle} rpm="6" radius={line.length} center={line.center} />,
    );

    expect(wrapper.instance().props.setLineAngle).toBeDefined();
    expect(wrapper.instance().props.setLineAngle).toEqual(setLineAngle);
  });

  it('should set an angle to start on', () => {
    const wrapper = shallow(
      <SpinLine show setLineAngle={setLineAngle} rpm="6" radius={line.length} center={line.center} />,
    );

    expect(wrapper.instance().state.angle).toEqual(0);
  });

})

describe('interval function', () => {

  it('should call the setLineAngle function', () => {
    const setLineAngle = jest.fn();
    const wrapper = shallow(
      <SpinLine show setLineAngle={setLineAngle} rpm="6" radius={line.length} center={line.center} />,
    );
    expect(setLineAngle).not.toBeCalled();

    jest.runOnlyPendingTimers();
    let newAngle = wrapper.instance().state.angle;

    expect(setLineAngle).toBeCalled();
    expect(setLineAngle).toBeCalledWith(newAngle);
  });

  it('should increase the angle of the line', () => {
    const setLineAngle = jest.fn();
    const wrapper = shallow(
      <SpinLine show setLineAngle={setLineAngle} rpm="6" radius={line.length} center={line.center} />,
    );

    let angle = wrapper.instance().state.angle;

    jest.runOnlyPendingTimers();

    expect(wrapper.instance().state.angle).toBeGreaterThan(angle);
  });

  it('should reset the angle of the line to zero', () => {
    const setLineAngle = jest.fn();
    const wrapper = shallow(
      <SpinLine show setLineAngle={setLineAngle} rpm="6" radius={line.length} center={line.center} />,
    );

    wrapper.instance().state.angle = 359.9;

    jest.runOnlyPendingTimers();

    expect(wrapper.instance().state.angle).toEqual(0);
  });

})

describe('tear down of component', () => {

  it('should clear the spinline interval', () => {
    const wrapper = shallow(
      <SpinLine show setLineAngle={setLineAngle} rpm="6" radius={line.length} center={line.center} />,
    );
    let spy = jest.spyOn(wrapper.instance(), 'componentWillUnmount')

    wrapper.instance().componentWillUnmount();

    expect(spy).toBeCalled();
    expect(clearInterval).toBeCalled();
  });

})




