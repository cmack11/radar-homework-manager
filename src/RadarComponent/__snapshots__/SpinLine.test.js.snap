// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initialization should initialize a null SpinLine 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SpinLine
    center={
      Object {
        "x": 10,
        "y": 15,
      }
    }
    radius={100}
    rpm="6"
    setLineAngle={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): null,
  Symbol(enzyme.__nodes__): Array [
    null,
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`initialization should initialize the SpinLine 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SpinLine
    center={
      Object {
        "x": 10,
        "y": 15,
      }
    }
    radius={100}
    rpm="6"
    setLineAngle={[Function]}
    show={true}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "id": "spinLine",
      "stroke": undefined,
      "strokeWidth": "10",
      "x1": 10,
      "x2": 10.000000000000005,
      "y1": 15,
      "y2": -85,
    },
    "ref": null,
    "rendered": null,
    "type": "line",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "id": "spinLine",
        "stroke": undefined,
        "strokeWidth": "10",
        "x1": 10,
        "x2": 10.000000000000005,
        "y1": 15,
        "y2": -85,
      },
      "ref": null,
      "rendered": null,
      "type": "line",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
